src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs

```diff
@@ -2583,85 +2583,136 @@ private T GetFieldValueInternal<T>(int i)
 
         private T GetFieldValueFromSqlBufferInternal<T>(SqlBuffer data, _SqlMetaData metaData)
         {
-            Type typeofT = typeof(T);
-            if (_typeofINullable.IsAssignableFrom(typeofT))
+            // this block of type specific shortcuts uses RyuJIT jit behaviours to achieve fast implementations of the primitive types
+            // RyuJIT will be able to determine at compilation time that the typeof(T)==typeof(<primitive>) options are constant
+            // and be able to remove all implementations which cannot be reached. this will eliminate non-specialized code for value types
+            Type dataType = data.GetTypeFromStorageType(false);
+            if (typeof(T) == typeof(int) && dataType == typeof(int))
             {
-                // If its a SQL Type or Nullable UDT
-                object rawValue = GetSqlValueFromSqlBufferInternal(data, metaData);
-
-                if (typeofT == s_typeofSqlString)
+                return data.Int32As<T>();
+            }
+            else if (typeof(T) == typeof(byte) && dataType == typeof(byte))
+            {
+                return data.ByteAs<T>();
+            }
+            else if (typeof(T) == typeof(short) && dataType == typeof(short))
+            {
+                return data.Int16As<T>();
+            }
+            else if (typeof(T) == typeof(long) && dataType == typeof(long))
+            {
+                return data.Int64As<T>();
+            }
+            else if (typeof(T) == typeof(bool) && dataType == typeof(bool))
+            {
+                return data.BooleanAs<T>();
+            }
+            else if (typeof(T) == typeof(double) && dataType == typeof(double))
+            {
+                return data.DoubleAs<T>();
+            }
+            else if (typeof(T) == typeof(float) && dataType == typeof(float))
+            {
+                return data.SingleAs<T>();
+            }
+            else if (typeof(T) == typeof(Guid) && dataType == typeof(Guid))
+            {
+                return (T)(object)data.Guid;
+            }
+            else if (typeof(T) == typeof(decimal) && dataType == typeof(decimal))
+            {
+                return (T)(object)data.Decimal;
+            }
+            else if (typeof(T) == typeof(DateTimeOffset) && dataType == typeof(DateTimeOffset) && _typeSystem > SqlConnectionString.TypeSystem.SQLServer2005 && !metaData.IsNewKatmaiDateTimeType)
```

Comment:
Why are we checking for  !metaData.IsNewKatmaiDateTimeType here?

Reply:
When I traced through how datetime was handled I found that there is special casing going on for old database versions and that it's checked using this flag. Katmai datetime values are really stored as strings [according to the code](https://github.com/dotnet/corefx/blob/d866aa6eb7f4f5c2df5892b09d694280d2982150/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBuffer.cs#L448) so if you want to get a datetime you have to know whether it's directly stored or has to go through reinterpretation. 

In that line particularly I took the logic used in [GetSqlValueFromSqlBufferInternal](https://github.com/dotnet/corefx/blob/d866aa6eb7f4f5c2df5892b09d694280d2982150/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs#L2526) and did the check to see if I can tell if I've really got a datetime I can easily return, if I have then do so, if not follow the old path to ensure compatibility.

Reply:
Thanks, now what would be the case if Sql server version is bigger than 2005 & datatype is datetimeoffset. 
Doesn't the check for metaData.IsNewKatmaiDateTimeType return true, hence the !metaData.IsNewKatmaiDateTimeType would eventually be false?

Reply:
You're right, I've inverted both parts when I only need to invert the version check. It should (and now does) check that the type system supports date/time/offset and that the column is one of those types. This would have forced dates down the compatibility path which wouldn't have been faster but also wouldn't be slower.

Reply:
Thanks. I will rerun the CI.